classification/models/objectives.py

```diff
@@ -0,0 +1,388 @@
+# Copyright 2017 Google Inc. and Skytruth Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import calendar
+from collections import namedtuple, OrderedDict
+import datetime
+import logging
+import numpy as np
+import tensorflow as tf
+import tensorflow.metrics as metrics
+from classification import metadata
+import pytz
+""" Terminology in the context of objectives.
+    
+    Net: the raw input to an objective function, an embeddeding that has not
+         yet been shaped for the predictive task in hand.
+    Logits: the input to a softmax classifier.
+    Prediction: the output of an objective function, be it class probabilities
+                from a categorical function, or a continuous output vector for
+                a regression.
+"""
+
+Trainer = namedtuple("Trainer", ["loss", "update_ops"])
+TrainNetInfo = namedtuple("TrainNetInfo", ["optimizer", "objective_trainers"])
+
+EPSILON = 1e-20
+
+
+def f1(recall, precision):
+    rval, rop = recall
+    pval, pop = precision
+    f1 = 2.0 / (1.0 / rval + 1.0 / pval)
+    return (f1, f1)
+
+
+class ObjectiveBase(object):
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, metadata_label, name, loss_weight, metrics):
+        """
+        args:
+            metadata_label:
+            name: name of this objective (for metrics)
+            loss_weight: weight of this objective (so that can increase decrease relative to other objectives)
+            metrics: which metrics to include. Options are currently ['all', 'minimal']
+
+        """
+        self.metadata_label = metadata_label
+        self.name = name
+        self.loss_weight = loss_weight
+        self.prediction = None
+        self.metrics = metrics
+
+    @abc.abstractmethod
+    def build(self, net):
+        pass
+
+    def create_metrics(self, labels):
+        raw_metrics = self.create_raw_metrics(labels)
+        try:
+            eval_metrics = {"{}/{}".format(self.name, k) : v for (k, v) in raw_metrics.items()}
+        except:
+            logging.warning("Problem creating eval_metrics in {}".format(self))
+            return {}
+        for k, v in eval_metrics.items():
+            tf.summary.scalar(k, v[1])
+        return eval_metrics
+
+
+
+class RegressionObjective(ObjectiveBase):
```

Comment:
I'm not sure if you still have the regression heads for length, tonnage etc. But if you do, I've had a lot of recent success with quantising regression outputs, using softmax and then training with cross-entropy (having quantised and then one-hot encoded the label). Then using the tenth to ninetieth (for instance) percentile probability gap gives a nice estimate of model uncertainty.

Reply:
@d40cht , interesting. I've toyed with that idea, but never got around to trying it. Do you ever end up with bimodal distributions? If you do, do you do anything clever about it?